lib/dest/write-contents/write-buffer.js

```diff
@@ -1,14 +1,30 @@
 'use strict';
 
+var iconv = require('iconv-lite');
+
 var fo = require('../../file-operations');
+var getCodec = require('../../codecs');
 
 function writeBuffer(file, optResolver, onWritten) {
+  var encoding = optResolver.resolve('encoding', file);
+  var codec = getCodec(encoding);
+  if (encoding && !codec) {
+    return onWritten(new Error('Unsupported encoding: ' + encoding));
+  }
+
   var opt = {
     mode: file.stat.mode,
     flag: optResolver.resolve('flag', file),
   };
 
-  fo.writeFile(file.path, file.contents, opt, onWriteFile);
+  var contents = file.contents;
+
+  if (encoding && codec.enc !== 'utf8') {
+    contents = iconv.decode(contents, 'utf8', { stripBOM: false });
```

Comment:
Yes, any characters that are not supported by the target encoding will be replaced with `'�'` (multi-byte encodings) or `'?'` (single-byte encodings).

https://github.com/ashtuchkin/iconv-lite/blob/master/lib/index.js#L15

That's just an inescapable consequence of different encodings including different characters, though. 

Having said that I believe that Unicode covers the vast majority of characters that were previously part of various older sets. For the vast majority of those that are not exactly replicated it has decent look-alikes / variants, and it should be the core business of something like iconv to leverage those instead.

For example, I am using an old Simplified Chinese set in the unit tests which predates Unicode, and that works out just fine (albeit for a short sentence).

Since UTF8 can reflect all of Unicode, it is certainly our best bet for internal/default encoding.

Of course Node buffers can contain other encodings, in fact it can contain raw binary stuff that's not encoded text at all. So if folks fill the buffer with non-UTF8 contents *and* then attempt to decode that to text, garbage will ensue. My contention is that that's not something vinyl-fs can, or should, protect against.

For what it's worth, it's still possible to read/write pure binary files with vinyl-fs. I took care to allow the `encoding` option to be `'false'` to do no transcoding at all (and also in that case it won't strip BOMs). I forgot to document that possibility, come to think of it, and I should also add separate tests for it.